WIP Child modules release

snopoke opened the pull request:
http://manage.dimagi.com/default.asp?171869

snopoke committed:
update data model

snopoke committed:
add tests for basic child modules

czue commented:
guessing you're aware but tests failed
```
======================================================================
FAIL: test_child_module_session_datums_added (corehq.apps.app_manager.tests.test_suite.BasicModuleAsChildTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/tests/test_suite.py", line 958, in test_child_module_session_datums_added
    self.assertXmlPartialEqual(self.get_xml('child-module-entry-datums-added'), self.app.create_suite(), "./entry")
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/tests/util.py", line 20, in assertXmlPartialEqual
    self.assertXmlEqual(expected, actual, normalize=False)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/tests/util.py", line 26, in assertXmlEqual
    _check_shared(expected, actual, LXMLOutputChecker(), "xml")
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/tests/util.py", line 113, in _check_shared
    raise AssertionError(message)
AssertionError: XML mismatch
--- want.xml
+++ got.xml
@@ -20,9 +20,8 @@
     <instance id="casedb" src="jr://instance/casedb"/>
     <instance id="commcaresession" src="jr://instance/session"/>
     <session>
-      <datum detail-select="m0_case_short" id="case_id" nodeset="instance('casedb')/casedb/case[@case_type='gold-fish'][@status='open']" value="./@case_id"/>
-      <datum function="uuid()" id="case_id_new_guppy_0"/>
-      <datum detail-confirm="m1_case_long" detail-select="m1_case_short" id="case_id_guppy" nodeset="instance('casedb')/casedb/case[@case_type='guppy'][@status='open'][index/parent=instance('commcaresession')/session/data/case_id]" value="./@case_id"/>
+      <datum detail-select="m0_case_short" id="parent_id" nodeset="instance('casedb')/casedb/case[@case_type='gold-fish'][@status='open']" value="./@case_id"/>
+      <datum detail-confirm="m1_case_long" detail-select="m1_case_short" id="case_id" nodeset="instance('casedb')/casedb/case[@case_type='guppy'][@status='open'][index/parent=instance('commcaresession')/session/data/parent_id]" value="./@case_id"/>
     </session>
   </entry>
 </partial>
----------------------------------------------------------------------
```

czue commented:
is there a spec/ticket for this?

snopoke commented:
@czue updated the description with the ticket. The tests are meant to fail since the functionality isn't implemented yet. I just made this change as a way to think  through what the necessary changes would be.

kaapstorm committed:
Merge branch 'master' into child_module

kaapstorm committed:
Child modules

kaapstorm commented:
Added child modules for normal modules. 

Still to do: usercase-only modules.

ping @snopoke, buddy @millerdev

kaapstorm commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1936,42 +1975,29 @@ def get_target_module(case_type, module_id, with_product_details=False):
-                        this_datum.id = that_datum.id
+                        avoid_duplicate_ids(action, this_datum, parent_datum, datum_ids, changed_ids_by_case_tag)
+                    elif module.module_type == 'basic':
+                        datums.insert(index, parent_datum_meta)
```
@snopoke, can you think of a scenario where we need to check for more than just module_type == 'basic'? This change caters for current tests, but I'm wondering whether I'm missing anything.

kaapstorm commented:
(I think I'll do the stuff for child modules of usercase-only modules in a different branch, so that Nick doesn't have to wait for it.)

kaapstorm committed:
Only insert parent datum if parent_select is active

kaapstorm committed:
Mimic advanced module datum id naming convention

millerdev commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1913,6 +1914,44 @@ def get_target_module(case_type, module_id, with_product_details=False):
+
+        return datums, assertions
+
+    def add_parent_datums(self, datums, module):
```
I thought *datum* was the singular form of *data* ([wikipedia agrees](https://en.wikipedia.org/wiki/Datum)). Or maybe these are [geodetic datums](http://oceanservice.noaa.gov/facts/datum.html)? Obviously no action required.

kaapstorm replied:
:smiley: `datums` bothered me too. But I left it the way I found it.

snopoke replied:
We use the term datum as a name of an element and not as the singular of
data, hence datums.
On 3 Jul 2015 18:15, "Norman Hooper" <notifications@github.com> wrote:

> In corehq/apps/app_manager/suite_xml.py
> <https://github.com/dimagi/commcare-hq/pull/7246#discussion_r33872801>:
>
> > @@ -1913,6 +1914,44 @@ def get_target_module(case_type, module_id, with_product_details=False):
> >              except IndexError:
> >                  pass
> >
> > +        self.add_parent_datums(datums, module)
> > +
> > +        return datums, assertions
> > +
> > +    def add_parent_datums(self, datums, module):
>
> [image: :smiley:] datums bothered me too. But I left it the way I found
> it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/7246/files#r33872801>.
>

millerdev commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1913,6 +1914,44 @@ def get_target_module(case_type, module_id, with_product_details=False):
+    def add_parent_datums(self, datums, module):
+
+        def update_refs(action_, datum_, changed_ids):
+            if action:
```
This is not referencing a local variable (although `action_` looks very similar). Is that intentional?

kaapstorm replied:
Well spotted. Thanks @millerdev.

kaapstorm committed:
Missing underscore

kaapstorm committed:
Guess case_id of submodule form

TODO: Don't guess! Case ID could have "_n" suffix if there is more than one submodule with the same case type.

kaapstorm committed:
Merge branch 'master' into child_module

kaapstorm committed:
Tear down mock

kaapstorm commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1943,6 +1944,44 @@ def get_target_module(case_type, module_id, with_product_details=False):
+                    i = 0
+                    while new_id in datum_ids_:
+                        new_id = '_'.join((parent_datum_.id, datum['case_type'], str(i)))
+                        i += 1
```
@snopoke, are lines 1973-1976 necessary? Surely there will only be one datum per case type for a form? Is that true for advanced forms too?

kaapstorm replied:
@snopoke, nvm, I see it's possible from the interface for advanced forms, so I'll leave it in.

snopoke replied:
For normal forms it definitely isn't necessary. It is possible with Advanced Forms but I don't think you need to worry about it since datums for advanced forms are named using the action tags already.

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1966,42 +2005,33 @@ def get_target_module(case_type, module_id, with_product_details=False):
+                        module.parent_select.active and
+                        module.parent_select.module_id == root_module.unique_id
+                    ):
+                        datums.insert(index, parent_datum_meta)
```
What is this last `elif` doing? Its not covered by any of the existing test cases.

kaapstorm replied:
Hmmm! When I was stepping through the code originally I must have thought that parent_datum_meta['requires_selection'] might be true for basic modules. But I must have been wrong, because it's working correctly without that `elif` for the example I was using. I'll take it out.

kaapstorm committed:
Case ID suffix explained

kaapstorm committed:
Remove unnecessary elif

kaapstorm committed:
Merge branch 'master' into child_module

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1943,6 +1944,46 @@ def get_target_module(case_type, module_id, with_product_details=False):
+                # Only advanced module actions have a parent_tag attribute. Basic modules don't need one
+                # because they only deal with one case type. Set parent_tag to None for them.
+                parent_tag = getattr(action_, 'parent_tag', None)
+                if parent_tag in changed_ids:
```
Nitpicking but maybe do `if parent_tag and parent_tag in changed_ids` to make it clear that you don't expect `None` or `''` to be a key in `changed_ids`

snopoke replied:
I see this isn't actually true. Deleting comment

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1943,6 +1944,46 @@ def get_target_module(case_type, module_id, with_product_details=False):
+
+        def avoid_duplicate_ids(action_, this_datum_, parent_datum_, datum_ids_, changed_ids):
+            if action_:
+                case_tag = getattr(action_, 'case_tag', None)  # Set case_tag to None for basic modules
```
I don't think this will ever happen in practice but it seems odd to have the same case_tag (`None`) for multiple actions since it ends up being a key in a dictionary.

I don't think necessary to change now since it would be a bit involved but just raising as a flag.

kaapstorm replied:
This should keep the original behaviour for advanced modules. For basic modules, I thought there would only be one action per form ... so we don't actually need a key at all. (Am I wrong about that?) But keeping a key of `None` lets us use the same code for both.

It is a bit obscure though. It might be cleaner and more obvious if we treated basic and advanced modules differently. What do you think?

snopoke replied:
Yes, your assumptions are correct its just that for someone who doesn't
know that this code doesn't make sense (or looks like it could be a bug). I
was also wondering if we can't just treat basic modules differently since
this whole algorithm seems overkill.

snopoke commented on corehq/apps/app_manager/templates/app_manager/module_view.html:
```diff
@@ -228,6 +228,20 @@
+                            <option value="{{module.unique_id}}">{{ module.name|trans:langs }}</option>
+                        {% endfor %}
+                        </select>
+                        <!-- need a js util to get module name by unique id -->
```
comment still relevant?

snopoke replied:
I see just copied but can probably delete.

kaapstorm replied:
Good call. I'm dropping it.

kaapstorm committed:
Dropped old comment

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1943,6 +1944,46 @@ def get_target_module(case_type, module_id, with_product_details=False):
+                if parent_datum_.id in datum_ids_:
+                    # We can't set a new ID to an ID we already have. We need to rename current ID
+                    datum = datum_ids_[parent_datum_.id]
+                    new_id = '_'.join((parent_datum_.id, datum['case_type']))
```
I think it makes more sense to use `datum.id` rather than `parent_datum_.id` here (and below). I know they have the same value but it makes more sense to create the new ID from the current ID rather than the parent's.

kaapstorm replied:
Do you mean `new_id = '_'.join((this_datum_.id, datum['case_type']))`? 

When I do that `new_id` is set to `parent_id_guppy` (for example) instead of `case_id_guppy`, because the datum in the parent entry has the id `case_id`, but the datum in this entry has the id `parent_id`.

snopoke replied:
No I mean `new_id = '_'.join((datum.id, datum['case_type']))`

snopoke commented on corehq/apps/app_manager/xform.py:
```diff
@@ -1276,6 +1276,15 @@ def make_delegation_stub_case_block():
                 )
                 if 'external_id' in actions['open_case'] and actions['open_case'].external_id:
                     extra_updates['external_id'] = actions['open_case'].external_id
+            elif module.parent_select.active:
```
This should be `elif module.root_module_id and module.parent_select.active:`

snopoke commented on corehq/apps/app_manager/xform.py:
```diff
@@ -1276,6 +1276,15 @@ def make_delegation_stub_case_block():
+            elif module.parent_select.active:
+                # This is a submodule. case_id will have changed to avoid a clash with the parent case.
+                # Case type is enough to ensure uniqueness for normal forms. No need to worry about a suffix.
+                case_id = '_'.join((CASE_ID, form.get_case_type()))
```
After all the work done in suite_xml.py this feels like cheating. OR is there a way we can shortcut the algorithm in suite_xml.py?

kaapstorm replied:
It is kinda cheating. This is true basic modules but not advanced modules (if I understand advanced modules correctly). If we treat basic and advanced modules differently in avoid_duplicate_ids() in suite_xml (of have a different function for each), then we can do this there. Might be more readable too. :-/

kaapstorm committed:
Add BASIC_CHILD_MODULE toggle

kaapstorm committed:
Exclude this module from parent module dropdown

kaapstorm committed:
Check root_module_id set

kaapstorm committed:
Simpler and more readable

kaapstorm commented:
Hopefully a little simpler and more readable now, @snopoke.

kaapstorm committed:
Revert "Simpler and more readable"

This reverts commit 3dd2c9b23254e59daff5c1cdfcaa63598cb75ba3.

snopoke commented:
OK, this is :white_check_mark: once build passes
